api-consumption/src/main/java/org/zalando/nakadi/service/subscription/state/StreamingState.java

```diff
@@ -370,6 +382,49 @@ private void streamToOutput(final boolean streamTimeoutReached) {
         }
     }
 
+    private void sendToDeadLetterQueue(final ConsumedEvent event, final int failedCommitsCount) {
+        // TODO: how do we handle AVRO here?
+
+        final String failedEventString = new String(event.getEvent(), StandardCharsets.UTF_8);
+        final JSONObject failedEvent = new JSONObject(failedEventString);
+
+        final JSONObject errorMessage =
+                new JSONObject()
+                .put("message", "skipped due to failed commits count: " + failedCommitsCount);
+
+        final SubscriptionCursorWithoutToken cursor =
+                getContext().getCursorConverter().convertToNoToken(event.getPosition());
+
+        final JSONObject deadLetter =
+                new JSONObject()
+                .put("subscription_id", getContext().getSubscription().getId())
+                .put("event_type", cursor.getEventType())
+                .put("partition", cursor.getPartition())
+                .put("offset", cursor.getOffset())
+                .put("error", errorMessage)
+                .put("event", failedEvent)
+                .put("metadata",
+                        new JSONObject()
+                        .put("eid", getContext().getUuidGenerator().randomUUID().toString())
+                        .put("occurred_at", Instant.now().toString()));
+
+        final JSONArray deadLetterBatch =
+                new JSONArray().put(deadLetter);
+
+        final EventPublishResult result;
+        try {
+            result = getContext().getEventPublisher().publishInternal(
```

Comment:
Does it retry publishing internally on failure?

Reply:
That depends on the configuration.  Currently — no.